File: go/test/protocol/test_case_catch.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
package protocol

import "gopkg.in/yaml.v3"

const (
	HeaderNameKey         = "header-name"
	HeaderValueKey        = "header-value"
	TimeoutNameKey        = "timeout-name"
	TimeoutValueKey       = "timeout-value"
	PropertyNameKey       = "property-name"
	PropertyValueKey      = "property-value"
	ProtocolVersionKey    = "protocol-version"
	SupportedProtocolsKey = "supported-protocols"
)

type testCaseCatch struct {
	ErrorKind    string            `yaml:"error-kind"`
	IsShallow    *bool             `yaml:"is-shallow"`
	IsRemote     *bool             `yaml:"is-remote"`
	Message      *string           `yaml:"message"`
	Supplemental map[string]string `yaml:"supplemental"`
}

type TestCaseCatch struct {
	testCaseCatch
}

func (catch *TestCaseCatch) UnmarshalYAML(node *yaml.Node) error {
	*catch = TestCaseCatch{}
	return node.Decode(&catch.testCaseCatch)
}

File: dotnet/test/Azure.Iot.Operations.Protocol.MetlTests/TestCaseCatch.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

namespace Azure.Iot.Operations.Protocol.MetlTests
{
    public class TestCaseCatch
    {
        public const string HeaderNameKey = "header-name";
        public const string HeaderValueKey = "header-value";
        public const string TimeoutNameKey = "timeout-name";
        public const string TimeoutValueKey = "timeout-value";
        public const string PropertyNameKey = "property-name";
        public const string PropertyValueKey = "property-value";
        public const string CommandNameKey = "command-name";
        public const string RequestProtocolKey = "request-protocol";
        public const string SupportedMajorProtocolVersions = "supported-protocols";

        public string? ErrorKind { get; set; }

        public bool? IsShallow { get; set; }

        public bool? IsRemote { get; set; }

        public object StatusCode { get; set; } = false;

        public string? Message { get; set; }

        public Dictionary<string, string?> Supplemental { get; set; } = new();

        public AkriMqttErrorKind GetErrorKind()
        {
            return ErrorKind switch
            {
                "missing header" => AkriMqttErrorKind.HeaderMissing,
                "invalid header" => AkriMqttErrorKind.HeaderInvalid,
                "invalid payload" => AkriMqttErrorKind.PayloadInvalid,
                "timeout" => AkriMqttErrorKind.Timeout,
                "cancellation" => AkriMqttErrorKind.Cancellation,
                "invalid configuration" => AkriMqttErrorKind.ConfigurationInvalid,
                "invalid state" => AkriMqttErrorKind.StateInvalid,
                "internal logic error" => AkriMqttErrorKind.InternalLogicError,
                "unknown error" => AkriMqttErrorKind.UnknownError,
                "execution error" => AkriMqttErrorKind.ExecutionException,
                "mqtt error" => AkriMqttErrorKind.MqttError,
                "unsupported version" => AkriMqttErrorKind.UnsupportedVersion,
                _ => throw new Exception($"unrecognized error kind string \"{ErrorKind}\""),
            };
        }
    }
}
